csrc/multidevice/utils.cpp

```diff
@@ -614,28 +558,202 @@ bool isResharding(const Expr* expr) {
+      sharded_id->isReduction()) {
+    // Device dimension, broadcast, and reduction do not affect allocation.
+    return true;
+  }
+
+  // This sharded logical ID may not be directly present in allocation domain.
+  // This indicates allocation domain has DID transformations.
+  std::optional<Layout> layout = canonicalizeLayout(tv);
+  if (!layout.has_value()) {
+    return false;
+  }
+
+  const std::vector<IterDomain*>& allocation_domain =
+      (*layout).allocation_domain;
+
+  NVF_ERROR(
+      std::is_permutation(
+          allocation_domain.begin(),
+          allocation_domain.end(),
+          tv->getLogicalDomain().begin(),
+          tv->getLogicalDomain().end()),
+      "The allocation domain ",
+      allocation_domain,
+      " is not a permutation of the logical domain ",
+      tv->getLogicalDomain());
+
+  // Check if sharded_id appears at the front.
+  size_t idx = 0;
+  for (IterDomain* id : allocation_domain) {
+    if (id == sharded_id) {
+      return idx == 0;
+    }
+    if (id->isDeviceDim() || id->isBroadcast() || id->isReduction()) {
+      continue;
+    }
+    idx++;
+  }
+  NVF_THROW(
+      "Should never reach here - sharded_id must be found in allocation domain");
+}
+
+int64_t posInDomain(
+    const std::vector<IterDomain*>& domain,
+    const IterDomain* id) {
+  auto pos = std::find(domain.begin(), domain.end(), id);
   NVF_ERROR(
-      ir_utils::isTvOp(expr),
-      "Non-tv op is not supported : ",
-      expr->toString());
-
-  for (auto input : ir_utils::filterByType<TensorView>(expr->inputs())) {
-    for (auto output : ir_utils::filterByType<TensorView>(expr->outputs())) {
-      auto [shard_additions, shard_deletions] =
-          getShardingChanges(input, output);
+      pos != domain.end(),
+      "Expected id ",
+      id->toString(),
+      " in domain ",
+      domain);
+  return std::distance(domain.begin(), pos);
+}
+
+std::optional<CommunicationInfo> getCommunicationInfo(Expr* expr) {
+  TensorView* producer = expr->inputs().at(0)->as<TensorView>();
+  TensorView* consumer = expr->outputs().at(0)->as<TensorView>();
+  bool has_sharding_change = false;
+  std::optional<CommunicationInfo> communication_info = std::nullopt;
+
+  auto update_communication_info = [&](CommunicationType type,
+                                       IterDomain* p_sharded_id,
+                                       IterDomain* c_sharded_id) {
+    NVF_ERROR(!has_sharding_change, "Expected at most one sharding change");
+    has_sharding_change = true;
+    communication_info = CommunicationInfo{type, p_sharded_id, c_sharded_id};
+  };
+
+  const auto pairwise_map = PairwiseLogicalDomainMap(producer, consumer);
+  const auto p2c_map = pairwise_map.mapProducerToConsumer();
+  const auto c2p_map = pairwise_map.mapConsumerToProducer();
+
+  // This ignores device dimensions on reduction axis.
+  auto producer_pt_to_did =
+      mapDeviceParallelTypeToId(producer->getLoopDomain());
```

Comment:
```suggestion
      mapParallelTypeToId(producer->getLoopDomain());
```
I understand mapDeviceParalleTypeToId only looks at DIDs. But, given Stream is coming up, we should probably extend it for all non-serial ParallelTypes. I'd do this as a separate PR **before** this PR.

Reply:
See PR: https://github.com/NVIDIA/Fuser/pull/4536